File: 프로그래머스/JAVA/LV3_단어변환.java
import java.util.*;

class Words{
    String word;
    int root;
}
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        int count = 0;
        Words w = new Words();
        Queue <Words> q = new LinkedList<Words>();
        int[] distance = new int[50];
        
        w.word = begin;
        w.root = 0;
        q.add(w);
        //System.out.println(q.peek().word+" "+q.peek().root);
        
        String here;
        int r_node;
        
        while(!q.isEmpty()){
            w = q.poll();
            here = w.word;
            r_node = w.root;
            //System.out.println(w.word+" "+r_node+" "+distance[r_node]);
            
            for(int i = 0; i<words.length; i++){
                if(begin.equals(words[i]))  continue;
                if(i!=0 && i == r_node)   continue;
                if(distance[i] != 0) continue;
                for(int j = 0; j<begin.length(); j++){
                    if(here.charAt(j) != words[i].charAt(j)){
                        count++;
                    }
                    if(count > 1)   break;
                }
                if(count <= 1){
                    w.word = words[i];
                    w.root = i;
                    q.add(w);
                    distance[i] = distance[r_node]+1;
                    System.out.println(w.word+" "+i+" "+distance[i]);        
                }
                if(target.equals(w.word))  return distance[r_node]+1;
                count = 0;
            }
        }
        return answer;
    }
}

File: 프로그래머스/C++/LV3_단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> graph[50];
int Distance[50];

bool link(string begin, string next, int N){
    int count = 0;
    for(int i = 0; i<N; i++){
        if(begin[i] != next[i])    count++;
        if(count>1) return false;
    }
    return true;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    fill_n(Distance,50,-1);
    
    int N = begin.size();
    int M = words.size();
    
    for(int i = 0; i<M; i++){
        for(int j = 0; j<M; j++){
            if(j==i)    continue;
            if(link(words[i],words[j],N))    graph[i].push_back(j);
        }
    }//graph만들기
    int start = 0;
    for(int i = 0; i<M; i++){
        if(target==words[i]){
            start = i;
            break;
        }
        if(i== M-1 && target != words[i])
            return 0;
    }//target을 시작점으로 만들기
    //bfs시작
    queue <int> q;
    Distance[start] = 0;
    q.push(start);
    while(!q.empty()){
        int here = q.front();
        q.pop();
        if(link(words[here],begin,N)){
            answer = Distance[here]+1;
            return answer;
        }   
        for(int i = 0; i<graph[here].size(); i++){
            int next = graph[here][i];
            if(Distance[next]==-1){
                q.push(next);
                Distance[next] = Distance[here]+1;
            }
        }
    }
    answer = 0;
    return answer;
}
